### ISSUE TITLE
Grpc client streaming CONSTANTLY dropping packets

### ISSUE THREAD

**Issue by AhmadChaker** [grpc/grpc#11110]
EDIT: This seems to be an issue with C# clients, if I use a C++ client the issue is not replicable.

I have a C# client and server. 
I am attempting to stream data from the client to the server using just 1 write (so like a regular non-streaming call). However I am obtaining lots of dropped packets.

My proto file is:

// Plexus service
service Test1 {
        rpc Send (stream Capsule) returns (google.protobuf.Empty) {}
}

message Capsule {
        string   topic = 1;
}

On the client side I have a method that does the following call:

            var capsule = new Capsule {Topic = "AAAAA"};

            using (var call = _client.Send())
            {
                await call.RequestStream.WriteAsync(capsule);
                await call.RequestStream.CompleteAsync();
            }

On the server side I have the following code:
        public override async Task<Empty> Send(IAsyncStreamReader<Capsule> requestStream, ServerCallContext context)
        {
            while (await requestStream.MoveNext())
            {
                var req = requestStream.Current;
            }

            return new Empty();
        }

For my calls (in the vast majority of instances), the inside of the while loop is never reached, indicating that there is no data. In fact when I look at the private variables of the requestStream I can see that the CancellationToken is set to true indicating that some cancellation is going on, this is despite no cancellation request being sent through.

I'm using the latest NuGet packages as of this date.

Does anyone know why this is occurring ? I've attached a minimally reproducible sample

[MinimalSample.zip](https://github.com/grpc/grpc/files/995198/MinimalSample.zip)

**Comment by jtattermusch** [comment]
What platform are you running on? What grpc version?
Can you provide some details on what is the error that's happening with the call? (problems likely printed to the console once you enable more logging).

Enabling more logging:
`GrpcEnvironment.SetLogger(new TextWriterLogger(Console.Error));`
and set environment variable `GRPC_VERBOSITY=debug`

**Comment by jtattermusch** [comment]
Can you please update the issue with more details as requested?

**Comment by AhmadChaker** [comment]
I am running this on Windows with Grpc version 1.2.0 and also 1.3.0.
I will attempt to run the tests again with that flag set.

**Comment by AhmadChaker** [comment]
Quite odd, on the server side, the following exception appears for every stream call:
![image](https://cloud.githubusercontent.com/assets/20714770/26571289/9894c392-4558-11e7-9c89-7eaf0e2db138.png)

**Comment by jtattermusch** [comment]
I've looked at your code again and here's the problem:

```
using (var call = _client.Send())
{
                await call.RequestStream.WriteAsync(capsule);
                await call.RequestStream.WriteAsync(capsule);
                await call.RequestStream.CompleteAsync();
}
```

What this means is you send a two messages, then close request stream, but you don't  await the result of the call (`var result = await call;` or `var result = await call.ResponseAsync`).
The semantics of call.Dispose()  (invoked by the `using` block) is to Cancel the call if it hasn't finished successfully yet (call is treated as finished if response/error has been received) - that explains the cancelled flag being set to true. The exception on the server side signals that call has already been cancelled (so sending a response cannot be successful).
gRPC is not "losing" any packets, the client cancels the call  before they can be delivered.

Fixed code:
```
using (var call = _client.Send())
{
                await call.RequestStream.WriteAsync(capsule);
                await call.RequestStream.WriteAsync(capsule);
                await call.RequestStream.CompleteAsync();
                var result = await call;
}
```

This doesn't look like a gRPC issue, so I'm going to close it now.
